opal/datatype/opal_datatype_copy.c

```diff
@@ -58,12 +58,15 @@
 static bool opal_datatype_is_accel(void *dest, const void *src) {
     int dev_id;
     uint64_t flags;
-    if (opal_accelerator.check_addr(dest, &dev_id, &flags)) {
+    if (0 < opal_accelerator.check_addr(dest, &dev_id, &flags)) {
```

Comment:
I think use error as host buffer is not intuitive. `check_attr` should not have returned error for host buffer in the first place

Reply:
Also can you point me to the comment regarding the return value of opal_accelerator.check_addr?

Reply:
Check addr returns an error because cuda contexts have not been initialized which is a result of delayed initialization.

Reply:
```
/**                                                                                                                     
 * Check whether a pointer belongs to an accelerator or not.                                                            
 * interfaces                                                                                                           
 *                                                                                                                      
 * @param[IN] addr           Pointer to check                                                                           
 * @param[OUT] dev_id        Returns the device id against which the memory was allocated                               
 *                           or MCA_ACCELERATOR_NO_DEVICE_ID                                                            
 * @param[OUT] flags         Set to 0 or an Accelerator flag to indicate additional                                     
 *                           information about the corresponding pointer                                                
 *                                                                                                                      
 * @retval <0                An error has occurred.                                                                     
 * @retval 0                 The buffer does not belong to a managed buffer                                             
 *                           in device memory.                                                                          
 * @retval >0                The buffer belongs to a managed buffer in                                                  
 *                           device memory.                                                                             
 */                                                                                                                     
typedef int (*opal_accelerator_base_module_check_addr_fn_t)(                                                            
    const void *addr, int *dev_id, uint64_t *flags);
```

Reply:
I thought I handled this case with this check though:

```
        /* If cuda is not initialized, assume it is a host buffer. */                                                   
        if (CUDA_ERROR_NOT_INITIALIZED == result) {                                                                     
            return 0;  
```

Reply:
Either way, I prefer this change anyway. If device buffer checking fails, it's also highly likely that a device copy will fail. The best we can do is fail as this is not a recoverable/ignorable failure and retries are also unlikely to succeed. Since the difference in a memcpy failing and forcing a failure here anyway is basically the same, I would prefer the chance for memcpy to succeed. This is also the original behavior before porting everything to the accelerator framework.

Reply:
Can you put these reasoning in the code as comment? It is counter intuitive to treat error as host buffer.

Reply:
I'm not sure how this fits into the discussion, but I personally advise not treating a function ("opal_accelerator.check_addr" in this case) that returns integers ranging from positive to negative as a "bool". Guaranteed to create confusion. The sign of the returned value has meaning, so why not just check for what it actually returns instead of confusing everyone?

Reply:
I'm having some difficulty parsing what the ask is here - Are you saying to turn this function (opal_datatype_is_accel) from a boolean to an int return? Or are you saying we should explicitly check the value of check_addr when it is negative and when it is 0?

Reply:
Perhaps I'm misreading something, but this function (`typedef int (*opal_accelerator_base_module_check_addr_fn_t)`) appears to return an `int`. But the test (`if (opal_accelerator.check_addr(dest, &dev_id, &flags))`) treats it as a `bool`, which from what I can tell generates confusion since the +/- values actually mean something. I've seen more confusion over the years when someone does that then I can count - hence my advice to test it as an `int` since that is what it actually is.

Reply:
Oh yes, the change in this commit treats the return as an int where it was a bool before, that's where the issue arose (IE. an error was considered the same as it being a device buffer, when in actuality it is almost certainly a host buffer).

Reply:
Sorry I'm not being entirely clear. First, I was supporting the change to test for `> 0` as being a very good one to do 😄 However, I personally would find it disturbing that "we can only assume it is a host buffer" when it appears you have more information that would make the assumption unnecessary - you just cut yourself off from accessing that info by defining the overall function as returning a `bool`. You inherently have a tri-state "sensor" - why chop off 1/3 of the information?

Reply:
The rationale is that if buffer detection fails, it is likely that the buffer is a host buffer and the assumption is that we will attempt the copy mechanism as if it is a host buffer. A failure here would be unrecoverable and we might as well try a regular memcpy and see if it works. Perhaps I poorly designed the accelerator check addr API as maybe it should have just been a boolean. I thought bubbling up an error would be useful but it really doesn't seem to be

Reply:
> it is likely that the buffer is a host buffer

Not the same statement as "definitely is" 

> and the assumption is that we will attempt the copy mechanism as if it is a host buffer.

What if you are wrong? Is there a consequence to such a mistake? Can it be avoided via your "check addr" function (perhaps that function needs to be improved to remove the ambiguity)?

Not being snarky here - I'm honestly asking because I don't know.

Reply:
The backend CUDA function used here is [`cudaPointerGetAttributes`](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__UNIFIED.html#group__CUDART__UNIFIED_1gd89830e17d399c064a2f3c3fa8bb4390), it only know about memory registered with the device. Thus if the pointer is not a GPU buffer, it is likely a host buffer or an incorrect buffer (but no distinction between the last two can be made).

Reply:
Ah, thanks for clarifying! Much appreciated.